solver/src/oneinch_solver.rs

```diff
@@ -4,3 +4,221 @@
+        Self {
+            settlement_contract,
+            client: Default::default(),
+        }
+    }
+
+    /// Settles a single sell order against a 1Inch swap.
+    async fn settle_order(&self, order: LimitOrder) -> Result<Settlement> {
+        debug_assert_eq!(
+            order.kind,
+            OrderKind::Sell,
+            "only sell orders should be passed to settle_order"
+        );
+
+        let spender = self.client.get_spender().await?;
+        let sell_token = ERC20::at(&self.web3(), order.sell_token);
+        let existing_allowance = sell_token
+            .allowance(self.settlement_contract.address(), spender.address)
+            .call()
+            .await?;
+
+        let swap = self
+            .client
+            .get_swap(SwapQuery {
+                from_token_address: order.sell_token,
+                to_token_address: order.buy_token,
+                amount: order.sell_amount,
+                from_address: self.settlement_contract.address(),
+                slippage: Slippage::basis_points(MAX_SLIPPAGE_BPS).unwrap(),
+                // Disable balance/allowance checks, as the settlement contract
+                // does not hold balances to traded tokens.
+                disable_estimate: Some(true),
+            })
+            .await?;
+
+        let mut settlement = Settlement::new(hashmap! {
+            order.sell_token => swap.to_token_amount,
+            order.buy_token => swap.from_token_amount,
+        });
+
+        settlement.with_liquidity(&order, order.sell_amount)?;
+
+        if existing_allowance < order.sell_amount {
+            settlement
+                .encoder
+                .append_to_execution_plan(Erc20ApproveInteraction {
+                    token: sell_token,
+                    spender: spender.address,
+                    amount: U256::MAX,
+                });
+        }
+        settlement.encoder.append_to_execution_plan(swap);
+
+        Ok(settlement)
+    }
+
+    fn web3(&self) -> DynWeb3 {
+        self.settlement_contract.raw_instance().web3()
+    }
+}
+
+impl Interaction for Swap {
+    fn encode(&self) -> Vec<EncodedInteraction> {
+        vec![(self.tx.to, self.tx.value, self.tx.data.clone())]
+    }
+}
+
+/// Maximum number of sell orders to consider for settlements.
+///
+/// This is mostly out of concern to avoid rate limiting and because 1Inch
+/// requests take a non-trivial amount of time.
+const MAX_SETTLEMENTS: usize = 5;
+
+#[async_trait::async_trait]
+impl Solver for OneInchSolver {
+    async fn solve(&self, liquidity: Vec<Liquidity>, _gas_price: f64) -> Result<Vec<Settlement>> {
+        let settlements = future::join_all(
+            liquidity
+                .into_iter()
+                .filter_map(|liquidity| match liquidity {
+                    Liquidity::Limit(order) if order.kind == OrderKind::Sell => Some(order),
+                    _ => None,
+                })
+                .take(MAX_SETTLEMENTS)
```

Comment:
It could be interesting to randomize which 5 orders get considered just in case there are some unsolvable orders that always happen to be at the start of the liquidity list.

Reply:
Added. I added the `rand` crate as a dependency (we were already pulling it in anyway).